modules/docs/src/06-guides/extract-request-info.md

```diff
@@ -0,0 +1,205 @@
+---
+sidebar_label: Extracting Request Info
+title: Extracting Request Information
+---
+
+There are times where the implementation of your route handlers may require more information from the underlying http4s request.
+You may want to store the raw request, check the request body against an HMAC value, pass along header values for tracing spans, etc.
+
+When possible, the implementation of things like this should be isolated to middleware and not passed into the route handlers themselves.
+However, there are times where this isn't possible. This guide shows you how to implement a middleware that uses `IOLocal` to pass the value
+of several headers from the request into the smithy4s route handlers.
+
+## What is IOLocal?
+
+[IOLocal](https://github.com/typelevel/cats-effect/blob/series/3.x/core/shared/src/main/scala/cats/effect/IOLocal.scala)
+is a construct that allows for sharing context across the scope of a `Fiber`. This means it allows you to get and set some value `A` in the `IOLocal`.
+This value will be accessible across the current `Fiber`. As a `Fiber` is forked into new fibers, the value of `A` is carried over to the new `Fiber`.
+However, the new `Fiber` will not be able to update the value kept on its parent or sibling fibers.
+
+This diagram, adapted from the [IOLocal docs](https://github.com/typelevel/cats-effect/blob/series/3.x/core/shared/src/main/scala/cats/effect/IOLocal.scala), illustrates this well:
+
+![IOLocal flow chart](/img/ioLocalDiagram.svg)
+
+## Example Implementation
+
+### Smithy Spec
+
+For this example, we are going to be working with the following smithy specification (taken from [smithy4s repo](https://github.com/disneystreaming/smithy4s/blob/main/sampleSpecs/hello.smithy)):
+
+```kotlin
+namespace smithy4s.hello
+
+use smithy4s.api#simpleRestJson
+
+@simpleRestJson
+service HelloWorldService {
+  version: "1.0.0",
+  // Indicates that all operations in `HelloWorldService`,
+  // here limited to Hello, can return `GenericServerError`.
+  errors: [GenericServerError],
+  operations: [Hello]
+}
+
+@error("server")
+@httpError(500)
+structure GenericServerError {
+  message: String
+}
+
+@http(method: "POST", uri: "/{name}", code: 200)
+operation Hello {
+  input: Person,
+  output: Greeting
+}
+
+
+structure Person {
+  @httpLabel
+  @required
+  name: String
+}
+
+structure Greeting {
+  @required
+  message: String
+}
+```
+
+See our [getting started documentation](../01-overview/01-intro.md) for instructions on how to use this specification to generate scala code.
+
+### Service Implementation
+
+Let's start by creating a case class that we will use to hold the value of some headers from our request.
+
+```scala mdoc:silent
+case class RequestInfo(contentType: Option[String], userAgent: Option[String])
```

Comment:
Instead of having `RequestInfo` hold options (though, I'm not sure quite whether the Http4s model forces the presence of those two headers), let's have `IOLocal` hold an `Option[Request[Info]]`, and for the service to receive an `IO[RequestInfo]`. In other words, populated by : 

```scala 
ioLocal.get.flatMap {
   case Some(value) => IO.pure(value)
   case None => IO.raiseError(new IllegalAccessException("Tried to access the value outside of the lifecycle of an http request") 
} 
```